### PR TITLE
NATIVE-577: Show matching category content in search

### PR DESCRIPTION
**Pull request by sarahsporck**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **NATIVE**.

### CODE REVIEW

**Comment by maxammann** [comment]
I fixed the title of this PR such that it matches the others.

File: src/modules/categories/components/CategoryList.js [hunk 1]
```diff
@@ -9,9 +9,16 @@ import styled from 'styled-components/native'
 import Image from '../../common/components/Image'
 import CategoryListCaption from '../../../modules/common/components/CategoryListCaption'
 
+export type SimpleModelType = {|
```
Comment (maxammann) [inline src/modules/categories/components/CategoryList.js:12]: Maybe `ListElementModelType` is a better name as there is no complex or advanced ModelType

File: src/modules/categories/components/ContentMatcher.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// @flow
```
Comment (maxammann) [inline src/modules/categories/components/ContentMatcher.js:1]: ContentMatcher is no component. Therefore it should be moved one level up.

File: src/routes/search/components/SearchModal.js [hunk 3]
```diff
@@ -51,27 +52,45 @@ class SearchModal extends React.Component<PropsType, StateType> {
...
+    const parser = new Parser({ ontext (data: string) { contentWithoutHtml.push(data) } })
     const categoriesWithContent = categoriesArray
-      .filter(category => !normalizeSearchString(category.title).includes(normalizedFilter))
-      .filter(category => normalizeSearchString(category.content).includes(normalizedFilter))
-      .sort((category1, category2) => category1.title.localeCompare(category2.title))
+      .filter(category => !normalizeSearchString(category.title).includes(normalizedFilter) && !category.isRoot())
+      .map((category: CategoryModel): ListEntryType => {
+        contentWithoutHtml = []
+        parser.write(category.content)
+        parser.end()
```
Comment (maxammann) [inline src/routes/search/components/SearchModal.js:76]: If you call parser.end() within the map function then I would assume that the parser also gets initialized here.

**Review by maxammann — APPROVED** [review]

File: release-notes/unreleased/NATIVE-577-CategorySearchWithContent.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+issue_key: NATIVE-577
+show_in_stores: true
+platforms:
+  - android
+  - ios
+en: The search now also highlights the content of a category it matches.
```
Comment (klinzo) [inline release-notes/unreleased/NATIVE-577-CategorySearchWithContent.yml:6]:
```suggestion
en: The search now also highlights the matching category content.
```

File: release-notes/unreleased/NATIVE-577-CategorySearchWithContent.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+de: Die Suche hebt nun auch den Inhalt hervor, wo sie mit dem Filter übereinstimmt.
```
Comment (klinzo) [inline release-notes/unreleased/NATIVE-577-CategorySearchWithContent.yml:7]:
```suggestion
de: Die Suche hebt nun auch den Inhalt hervor, der mit dem Filter übereinstimmt.
```

File: src/modules/categories/components/CategoryList.js (hunk 1 above)
Comment (klinzo) [reply src/modules/categories/components/CategoryList.js:12]: Kind of... its a simpler CategoryType with additional `contentWithoutHtml`, but I'd also vote for something different. I don't think `Element` and `Model` are necessary, maybe simply `CategoryListModelType?

File: src/modules/categories/components/ContentMatcher.js (hunk 2 above)
Comment (klinzo) [reply src/modules/categories/components/ContentMatcher.js:1]: We also usually don't have classes for functionality, do we? Of course this is also okay, just not sure if a class makes sense here.

File: src/modules/categories/components/CategoryListItem.js [hunk 6]
```diff
@@ -27,16 +31,17 @@ const DirectionContainer: StyledComponent<DirectionContainerPropsType, ThemeType
   flex-direction: ${props => contentDirection(props.language)};
 `
 
-const CategoryTitleContainer: StyledComponent<DirectionContainerPropsType, ThemeType, *> = styled.View`
+const CategoryEntryContainer: StyledComponent<DirectionContainerPropsType, ThemeType, *> = styled.View`
   flex: 1;
-  flex-direction: ${props => contentDirection(props.language)};
+  flex-direction: column;
```
Comment (klinzo) [inline src/modules/categories/components/CategoryListItem.js:36]: Why did you change this and moved it to `CategoryTitle`?

File: src/modules/categories/components/CategoryListItem.js [hunk 7]
```diff
@@ -78,13 +85,33 @@ class CategoryListItem extends React.Component<PropsType> {
     )
   }
 
+  getMatchedContent (numWordsSurrounding: number): ?Highlighter {
+    const { query, theme, category } = this.props
+    const textToHighlight = this.contentMatcher.getMatchedContent(query, category.contentWithoutHtml, numWordsSurrounding)
```
Comment (klinzo) [inline src/modules/categories/components/CategoryListItem.js:90]: Any reason why you have to initialize contentMatcher for the whole component?

File: src/modules/categories/components/CategoryListItem.js (hunk 6 above)
Comment (klinzo) [reply src/modules/categories/components/CategoryListItem.js:36]: Especially, is the matched text in the right direction?

File: src/modules/categories/components/ContentMatcher.js (hunk 2 above)
Comment (klinzo) [reply src/modules/categories/components/ContentMatcher.js:1]: This could also not really category specific but could be used for events, ... Maybe it should even live somewhere else, idk.

File: src/routes/search/components/SearchModal.js [hunk 8]
```diff
@@ -51,27 +52,45 @@ class SearchModal extends React.Component<PropsType, StateType> {
...
+          path: category.path
+        },
+        subCategories: []
+      })
+
+      )
+      .sort((category1, category2) => category1.model.title.localeCompare(category2.model.title))
 
     // find all categories whose contents but not titles include the filter text and sort them lexicographically
+    let contentWithoutHtml = []
```
Comment (klinzo) [inline src/routes/search/components/SearchModal.js:69]: Can't we inline this and convert to const?

File: src/modules/categories/components/ContentMatcher.js (hunk 2 above)
Comment (maxammann) [reply src/modules/categories/components/ContentMatcher.js:1]: Yeah the function could just be exported as `export const ...`

**Review by lulo4907 — APPROVED** [review]

File: src/modules/categories/components/CategoryListItem.js (hunk 7 above)
Comment (Taggotty) [reply src/modules/categories/components/CategoryListItem.js:90]: The main reason why I did it, is because that is what I did in the webapp and nobody complained. Furthermore it kind of feels incorrect to reinstantiate the contentMatcher on every render

File: src/modules/categories/components/ContentMatcher.js (hunk 2 above)
Comment (Taggotty) [reply src/modules/categories/components/ContentMatcher.js:1]: I think it was @Schedulaar who recommended using a separate class for the ContentMatcher in the webapp, as it simplifies testing. 



> Yeah the function could just be exported as `export const ...`

Do you mean that I scrap the class and keep the functions as local functions? My question then would be, whether it is possible to to test the functions the and whether I should apply those changes to the webapp.

File: src/routes/search/components/SearchModal.js (hunk 3 above)
Comment (Taggotty) [reply src/routes/search/components/SearchModal.js:75]: I am kind of worried that the Parser initialization will impact the performance at least visibly. But I will test it shortly.

Comment (Taggotty) [reply src/routes/search/components/SearchModal.js:75]: Seems to not matter

**Review by steffenkleinle — APPROVED** [review]
